File: app/src/main/java/lt/zilinskas/marius/testio/fragments/LoginFragment.java
package lt.zilinskas.marius.testio.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.jakewharton.retrofit2.adapter.rxjava2.HttpException;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import lombok.Getter;
import lombok.Setter;
import lt.zilinskas.marius.testio.R;
import lt.zilinskas.marius.testio.api.entities.Token;
import lt.zilinskas.marius.testio.api.entities.UserInfo;
import lt.zilinskas.marius.testio.utils.StringUtils;

public class LoginFragment extends BaseFragment {

    private static final String SAVED_USERNAME = "username";
    private static final String SAVED_PASSWORD = "password";

    private CustomViewHolder customViewHolder;
    private CompositeDisposable compositeDisposable;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        compositeDisposable = new CompositeDisposable();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login, container, false);
        customViewHolder = new CustomViewHolder(view, savedInstanceState);

        return view;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putString(SAVED_USERNAME, customViewHolder.username.getText().toString());
        outState.putString(SAVED_PASSWORD, customViewHolder.username.getText().toString());
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (compositeDisposable != null && !compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
    }

    private void login() {
        String username = customViewHolder.username.getText().toString();
        String password = customViewHolder.password.getText().toString();

        if (StringUtils.isEmpty(username) || StringUtils.isEmpty(password)) {
            Toast.makeText(getContext(), getString(R.string.enter_username_and_password), Toast.LENGTH_LONG).show();
            return;
        }

        Disposable disposable = testioApplication.getPlaygroundApi()
                .authorize(new UserInfo(username, password))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(getObserver());
        compositeDisposable.add(disposable);
    }

    private DisposableObserver<Token> getObserver() {
        return new DisposableObserver<Token>() {
            @Override
            public void onNext(Token token) {
                testioApplication.getSharedPreferences().setToken(token);
                startLoadingFragment();
            }

            @Override
            public void onError(Throwable e) {
                if (e instanceof HttpException && ((HttpException) e).code() == 401) {
                    Toast.makeText(getContext(), getString(R.string.incorrect_username_or_password), Toast.LENGTH_LONG).show();
                } else {
                    Toast.makeText(getContext(), getString(R.string.unable_to_login), Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onComplete() {
            }
        };
    }

    private void startLoadingFragment() {
        if (getActivity() == null) {
            return;
        }

        getActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.fragment_container, new LoadingFragment(), LoadingFragment.class.getName())
                .commit();
    }


    @Getter
    @Setter
    private class CustomViewHolder {
        private View parent;
        private EditText username;
        private EditText password;
        private Button login;

        CustomViewHolder(@NonNull View parent, @Nullable Bundle savedInstanceSate) {
            this.parent = parent;
            initViews();
            initListeners();
            restoreState(savedInstanceSate);
        }

        private void initViews() {
            username = parent.findViewById(R.id.usernameEditText);
            password = parent.findViewById(R.id.passwordEditText);
            login = parent.findViewById(R.id.loginButton);
        }

        private void initListeners() {
            login.setOnClickListener(v -> login());
        }

        private void restoreState(Bundle savedInstanceState) {
            if (savedInstanceState == null) {
                return;
            }

            username.setText(savedInstanceState.getString(SAVED_USERNAME, ""));
            password.setText(savedInstanceState.getString(SAVED_PASSWORD, ""));
        }
    }
}

File: app/src/main/java/com/giedrius/androidparty/task/view/LoginFragment.kt
package com.giedrius.androidparty.task.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.giedrius.androidparty.R
import com.giedrius.androidparty.task.navigation.NavigationListener
import com.giedrius.androidparty.task.utils.ApiListener
import com.giedrius.androidparty.task.server.login.LoginOutcome
import com.giedrius.androidparty.task.viewmodel.LoginViewModel
import dagger.android.support.AndroidSupportInjection
import dagger.android.support.DaggerFragment
import kotlinx.android.synthetic.main.fragment_login.*
import kotlinx.android.synthetic.main.fragment_login.view.*
import javax.inject.Inject

class LoginFragment(val navListener: NavigationListener) : DaggerFragment() {
    @Inject lateinit var loginViewModel: LoginViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AndroidSupportInjection.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = LayoutInflater.from(context).inflate(R.layout.fragment_login, container, false)
        view.login_button.setOnClickListener {
            login_progress_bar.visibility = View.VISIBLE
            login_text.visibility = View.VISIBLE
            sendCredentials(view.user_name_edit_text.text.toString(), view.password_edit_text.text.toString())
        }
        return view
    }

    private fun sendCredentials(username: String, password: String) {
        loginViewModel.login(object : ApiListener<LoginOutcome> {
            override fun <T> onResult(data: T) {
                when (data as LoginOutcome) {
                    LoginOutcome.SUCCESSFUL -> { navListener.onLoginPerformed() }
                    LoginOutcome.UNSUCCESSFUL -> { Toast.makeText(activity, "Login unsuccessful", Toast.LENGTH_LONG).show() }
                }
                login_text.visibility = View.GONE
                login_progress_bar.visibility = View.GONE
            }
        }, username, password)
    }

    companion object {
        @JvmStatic fun newInstance(listener: NavigationListener) = LoginFragment(listener).apply {}
    }
}
